### PR TITLE
Attempted fix for #1107

### PR DESCRIPTION
**Pull request by adqm**
I believe this should fix the broken behavior when an attempted `cd ..` would move the user into a nonexistent directory, and generally restructures a few things about `cd`.

### CODE REVIEW

File: xonsh/dirstack.py [hunk 1]
```diff
@@ -5,28 +5,45 @@
...
 
+
 def _get_cwd():
     try:
         return os.getcwd()
     except (OSError, FileNotFoundError):
         return None
 
 
+def _splitpath(path, sofar=tuple()):
```
Comment (scopatz) [inline xonsh/dirstack.py]: I think you can just do `sofar=()` for an empty tuple.

Comment (scopatz) [inline xonsh/dirstack.py]: And is in fact much faster, though it probably doesn't matter:

```console
scopatz@artemis ~ $ timeit "()"
100000000 loops, best of 3: 10.1 ns per loop
scopatz@artemis ~ $ timeit "tuple()"
10000000 loops, best of 3: 54.4 ns per loop
```

Comment (adqm) [inline xonsh/dirstack.py]: better yet; i can remove this function.  i wrote it for a previous version, but it's not being used in this code.

File: xonsh/dirstack.py [hunk 2]
```diff
@@ -5,9 +5,12 @@
 from glob import iglob
 from argparse import ArgumentParser
 
+from xonsh.tools import get_sep
+
 DIRSTACK = []
 """A list containing the currently remembered directories."""
 
+
 def _get_cwd():
```
Comment (scopatz) [inline xonsh/dirstack.py]: It seems like this function is not used anymore.  Mind killing it completely?

**Comment by scopatz** [comment]
This looks good to me!

File: xonsh/dirstack.py (hunk 2 above)
Comment (adqm) [inline xonsh/dirstack.py]: It looks like it is still used a few other places via imports, so I don't think we should kill it.

Comment (scopatz) [inline xonsh/dirstack.py]: OK

**Pull request merged**